### ISSUE TITLE
langchain agents

### ISSUE THREAD

**Issue by fsboehme** [dennishansen/iga#2]
just came across this super helpful [video](https://youtu.be/2xxziIWmaSA) – it gets into langchain agents at the end. that's probably the major missing piece here. (some of the other parts should help as well! embeddings!)

i was playing with iga a bit yesterday and kept running into the issue that it forgets or just doesn't stick to its instructions and reverts into thinking it can't modify files etc.

**Comment by dennishansen** [comment]
Just pushed some system instructions that help with remembering it can edit. But yeah- seems like langchain gives us a lot in terms of memory. Been meaning to look into it so I'll do so, thanks!

**Comment by dennishansen** [comment]
Gave it more explicit read write abilities and improved the prompt so may be worth trying again! Been looking into langchain/pinecone and it likely makes sense to integrate

**Comment by fsboehme** [comment]
Yeah, I played with it again briefly yesterday and it did a pretty good job! 

Having spent a little more time exploring langchain agents, I'm not as sure anymore that they'll add _that_ much. I think they basically work the same way under the hood. It might just abstract away some of the manual dev work and add a little more structure.

Definitely think pinecone or some other vector store is needed.

A couple other projects I've found you might also find interesting:

1. https://github.com/lightaime/camel
2. https://github.com/biobootloader/wolverine

**Comment by dennishansen** [comment]
These are great. 100% agree. Yeah i've been waiting for langchain to become needed for a particular task, and evaluating it then v pinecode or something else. And i think the multi-agent CAMEL type model will certainly make sense at some point. Also, the wolverine is a great example of chaining to fix code changes- will definitely be thinking about self-healing.

**Comment by renatokuipers** [comment]
What a beauty of a program. Its simple, easy to understand and AWESOME.
I played with it a lot and came to the conclusion that it missed some things here and there.
I added simplistic localised short-term and persistent localised long-term memory that is being utilized by the program. I also added a feature add for browsing wikipedia pages. (this still is a WIP though, since the output is not what I want....)

The problem is, I can't commit anything, so I just forked your project if its alright :)
